src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.cpp

```diff
@@ -0,0 +1,785 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "CommodityTariffConsts.h"
+#include "commodity-tariff-server.h"
+#include <cassert>
+#include <cstdint>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::Platform;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Globals;
+using namespace chip::app::Clusters::Globals::Structs;
+using namespace chip::app::Clusters::CommodityTariff;
+using namespace chip::app::Clusters::CommodityTariff::Structs;
+using namespace chip::app::CommodityTariffAttrsDataMgmt;
+using namespace CommodityTariffConsts;
+
+// static constexpr size_t kAuxSwitchesSettingsMax = 8;
+
+#define VerifyOrReturnError_LogSend(expr, code, ...)                                                                               \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        if (!(expr))                                                                                                               \
+        {                                                                                                                          \
+            ChipLogError(NotSpecified, __VA_ARGS__);                                                                               \
+            VerifyOrReturnError(expr, code);                                                                                       \
+        }                                                                                                                          \
+    } while (false)
+
+namespace CommonUtilities {
+static bool HasFeatureInCtx(TariffUpdateCtx * aCtx, Feature aFeature)
+{
+    return aCtx->mFeature.Has(aFeature);
+}
+
+static void CleanUpIDs(DataModel::List<const uint32_t> & IDs)
+{
+    if (!IDs.empty() && IDs.data())
+    {
+        MemoryFree(const_cast<uint32_t *>(IDs.data()));
+        IDs = DataModel::List<const uint32_t>();
+    }
+}
+
+static bool HasDuplicateIDs(const DataModel::List<const uint32_t> & IDs, std::unordered_set<uint32_t> & seen)
+{
+    for (auto id : IDs)
+    {
+        if (!seen.insert(id).second)
+        {
+            return true; // Duplicate found
+        }
+    }
+    return false;
+}
+
+/**
+ * @brief The comparator for specific struct types wrapped to Optional<Struct>
```

Comment:
This needs to document why it exists (i.e. why just using `Optional<T>::operator==` does not work here).